File: packages/gheim-py/src/gheim/openai/_audio.py
"""Audio endpoints — speech (TTS) is anonymize-only; transcriptions/translations
deanonymize the returned text (and SSE deltas if streaming)."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated, Any

from ..core.session import Session
from ..core.stream import StreamDeanonymizer
from ..detectors import Detector
from ._base import (
    AsyncStreamingResponse,
    StreamingResponse,
    anonymize_value,
    resolve_call_session,
)


# ---------- audio.speech (TTS) — anonymize-only ----------

@dataclass
class SpeechProxy:
    inner: Annotated[Any, "openai client's audio.speech object."]
    detector: Annotated[Detector, "Default detector for this client."]

    def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        for k in ("input", "instructions"):
            if k in kwargs:
                kwargs[k] = anonymize_value(kwargs[k], sess, detector)
        return self.inner.create(*args, **kwargs)


@dataclass
class AsyncSpeechProxy:
    inner: Annotated[Any, "openai async client's audio.speech object."]
    detector: Annotated[Detector, "Default detector for this client."]

    async def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        for k in ("input", "instructions"):
            if k in kwargs:
                kwargs[k] = anonymize_value(kwargs[k], sess, detector)
        return await self.inner.create(*args, **kwargs)


# ---------- audio.transcriptions / audio.translations — deanonymize text out ----------

def _restore_text_field(response: Any, session: Session) -> None:
    text = getattr(response, "text", None)
    if isinstance(text, str):
        response.text = session.restore(text)


def _patch_transcription_chunk(chunk: Any, buf: StreamDeanonymizer) -> None:
    # transformers.js / openai SSE shapes for transcription deltas can vary; handle
    # the documented ones. Skip events with no text payload.
    delta = getattr(chunk, "delta", None)
    if isinstance(delta, str):
        chunk.delta = buf.feed(delta)
        return
    text = getattr(chunk, "text", None)
    if isinstance(text, str):
        chunk.text = buf.feed(text)


@dataclass
class TranscriptionsProxy:
    inner: Annotated[Any, "openai client's audio.transcriptions object."]
    detector: Annotated[Detector, "Default detector for this client."]

    def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        if "prompt" in kwargs and isinstance(kwargs["prompt"], str):
            kwargs["prompt"] = anonymize_value(kwargs["prompt"], sess, detector)
        stream = bool(kwargs.get("stream"))
        response = self.inner.create(*args, **kwargs)
        if stream:
            return StreamingResponse(
                stream=response,
                session=sess,
                mutate_chunk=_patch_transcription_chunk,
            )
        _restore_text_field(response, sess)
        return response


@dataclass
class AsyncTranscriptionsProxy:
    inner: Annotated[Any, "openai async client's audio.transcriptions object."]
    detector: Annotated[Detector, "Default detector for this client."]

    async def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        if "prompt" in kwargs and isinstance(kwargs["prompt"], str):
            kwargs["prompt"] = anonymize_value(kwargs["prompt"], sess, detector)
        stream = bool(kwargs.get("stream"))
        response = await self.inner.create(*args, **kwargs)
        if stream:
            return AsyncStreamingResponse(
                stream=response,
                session=sess,
                mutate_chunk=_patch_transcription_chunk,
            )
        _restore_text_field(response, sess)
        return response


@dataclass
class TranslationsProxy:
    """Same shape as TranscriptionsProxy but no streaming support upstream."""

    inner: Annotated[Any, "openai client's audio.translations object."]
    detector: Annotated[Detector, "Default detector for this client."]

    def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        if "prompt" in kwargs and isinstance(kwargs["prompt"], str):
            kwargs["prompt"] = anonymize_value(kwargs["prompt"], sess, detector)
        response = self.inner.create(*args, **kwargs)
        _restore_text_field(response, sess)
        return response


@dataclass
class AsyncTranslationsProxy:
    inner: Annotated[Any, "openai async client's audio.translations object."]
    detector: Annotated[Detector, "Default detector for this client."]

    async def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        if "prompt" in kwargs and isinstance(kwargs["prompt"], str):
            kwargs["prompt"] = anonymize_value(kwargs["prompt"], sess, detector)
        response = await self.inner.create(*args, **kwargs)
        _restore_text_field(response, sess)
        return response


# ---------- AudioProxy: aggregator that exposes .speech, .transcriptions, .translations ----------

@dataclass
class AudioProxy:
    inner: Annotated[Any, "openai client's audio object."]
    detector: Annotated[Detector, "Default detector for this client."]

    def __post_init__(self) -> None:
        self.speech = SpeechProxy(inner=self.inner.speech, detector=self.detector)
        self.transcriptions = TranscriptionsProxy(
            inner=self.inner.transcriptions, detector=self.detector
        )
        self.translations = TranslationsProxy(
            inner=self.inner.translations, detector=self.detector
        )

    def __getattr__(self, name: str) -> Any:
        return getattr(self.inner, name)


@dataclass
class AsyncAudioProxy:
    inner: Annotated[Any, "openai async client's audio object."]
    detector: Annotated[Detector, "Default detector for this client."]

    def __post_init__(self) -> None:
        self.speech = AsyncSpeechProxy(inner=self.inner.speech, detector=self.detector)
        self.transcriptions = AsyncTranscriptionsProxy(
            inner=self.inner.transcriptions, detector=self.detector
        )
        self.translations = AsyncTranslationsProxy(
            inner=self.inner.translations, detector=self.detector
        )

    def __getattr__(self, name: str) -> Any:
        return getattr(self.inner, name)

File: packages/gheim-js/src/openai/_audio.ts
/** Audio endpoints — speech (anonymize-only); transcriptions/translations (round-trip). */
import type OpenAIClass from "openai";
import { Session } from "../core/session.ts";
import { StreamDeanonymizer } from "../core/stream.ts";
import type { Detector } from "../detectors/base.ts";
import { StreamingResponse, anonymizeValue, type CallExtras, resolveCallSession } from "./_base.ts";

interface AudioInner {
  speech: { create(body: unknown): Promise<unknown> };
  transcriptions: { create(body: unknown): Promise<unknown> };
  translations: { create(body: unknown): Promise<unknown> };
}

interface TranscriptionResp {
  text?: string;
}

interface TranscriptionChunk {
  delta?: string;
  text?: string;
}

function patchTranscriptionChunk(chunk: TranscriptionChunk, buf: StreamDeanonymizer): void {
  if (typeof chunk.delta === "string") {
    chunk.delta = buf.feed(chunk.delta);
    return;
  }
  if (typeof chunk.text === "string") {
    chunk.text = buf.feed(chunk.text);
  }
}

export class GheimSpeech {
  constructor(
    private readonly innerP: Promise<AudioInner>,
    private readonly defaultDet: Detector,
  ) {}

  async create(body: Record<string, unknown> & CallExtras): Promise<unknown> {
    const { session, detector, rest } = resolveCallSession(body, this.defaultDet);
    for (const k of ["input", "instructions"]) {
      if (k in rest) {
        rest[k] = await anonymizeValue(rest[k], session, detector);
      }
    }
    const inner = await this.innerP;
    return inner.speech.create(rest);
  }
}

export class GheimTranscriptions {
  constructor(
    private readonly innerP: Promise<AudioInner>,
    private readonly defaultDet: Detector,
  ) {}

  async create(body: Record<string, unknown> & CallExtras): Promise<unknown> {
    const { session, detector, rest } = resolveCallSession(body, this.defaultDet);
    if (typeof rest.prompt === "string") {
      rest.prompt = await anonymizeValue(rest.prompt, session, detector);
    }
    const inner = await this.innerP;
    const response = await inner.transcriptions.create(rest);
    if (rest.stream) {
      return new StreamingResponse(
        response as AsyncIterable<TranscriptionChunk>,
        session,
        patchTranscriptionChunk,
      );
    }
    const r = response as TranscriptionResp;
    if (typeof r.text === "string") r.text = session.restore(r.text);
    return response;
  }
}

export class GheimTranslations {
  constructor(
    private readonly innerP: Promise<AudioInner>,
    private readonly defaultDet: Detector,
  ) {}

  async create(body: Record<string, unknown> & CallExtras): Promise<unknown> {
    const { session, detector, rest } = resolveCallSession(body, this.defaultDet);
    if (typeof rest.prompt === "string") {
      rest.prompt = await anonymizeValue(rest.prompt, session, detector);
    }
    const inner = await this.innerP;
    const response = await inner.translations.create(rest);
    const r = response as TranscriptionResp;
    if (typeof r.text === "string") r.text = session.restore(r.text);
    return response;
  }
}

export class GheimAudio {
  readonly speech: GheimSpeech;
  readonly transcriptions: GheimTranscriptions;
  readonly translations: GheimTranslations;

  constructor(clientPromise: Promise<OpenAIClass>, detector: Detector) {
    const innerP = clientPromise.then((c) => c.audio as unknown as AudioInner);
    this.speech = new GheimSpeech(innerP, detector);
    this.transcriptions = new GheimTranscriptions(innerP, detector);
    this.translations = new GheimTranslations(innerP, detector);
  }
}
